src/snowflake/snowpark/_internal/ast_utils.py

```diff
@@ -183,29 +183,101 @@ def build_proto_from_struct_type(
         ast_field.nullable = field.nullable
 
```

Comment:
this may be orthogonal to this PR potentially, but I am curious: How do you think about supporting decorators?

I.e.
```
f = udf(lambda x: x + 1, return_type=IntegerType())

@udf(return_type=StringType())
def foo(s):
  return f'Hello {s}'
```

Could we model those with the current changes as well, or how would additional work look like?

This does not need to be part of this PR, I am just curious whether the changes here could help support this use case as well or what we'd need to adapt.

Reply:
This would be a good discussion topic. I haven't put any thought into it yet.

If I had to do this from scratch, my steps would be:

* Find out how both examples capture the functions for Snowflake execution. I assume some degree of Python reflection is somehow involved.
* In both examples, actually serializing the function is somewhat orthogonal (or abstracted away) from getting the AST.
* In both examples, the callable object is a Python object. The `@udf` decorator gets the function as a parameter, just like the `udf` call. Assuming a black-box abstraction that can serialize these callable objects, what is non-trivial about doing the work in a decorator?